scripts/browserstack.py

```diff
@@ -524,19 +562,28 @@ def getFailedTestsForBuild(buildId):
 
             exitStatus = 1
             testClassesToRun = getAllTestClassNames()
-            while numRetries >= 0:
-                testResults = runTests(
-                    appUrl, testUrl, args.is_nightly, testClassesToRun
-                )
-                print("-----------------")
-                exitStatus = testResults["exitStatus"]
-                updateObservabilityWithResults(testResults["buildId"])
-                if exitStatus == 0:
-                    break
-                else:
+            testResults = runTests(
+                appUrl, testUrl, args.is_nightly, testClassesToRun
+            )
+            print("-----------------")
+            exitStatus = testResults["exitStatus"]
+            updateObservabilityWithResults(testResults["buildId"])
+            if exitStatus != 0 and numRetries > 0:
+                os.environ["BROWSERSTACK_RERUN"] = "true"
+                print(f"Getting failed tests for build {testResults['buildId']}")
+                failedTestsDictionary = getFailedTestsForBuild(testResults["buildId"])
+                while numRetries > 0:
+                    updatedFailedTestsDictionary = {}
                     numRetries -= 1
-                    testClassesToRun = getFailedTestsForBuild(testResults["buildId"])
-                    os.environ["BROWSERSTACK_RERUN"] = "true"
+                    for failedDevice in failedTestsDictionary:
+                        testResults = runTestsForFailure(appUrl, testUrl, failedDevice, failedTestsDictionary[failedDevice])
+                        exitStatus = testResults["exitStatus"]
+                        if exitStatus != 0 and numRetries > 0:
+                            updatedFailedTestsDictionary[failedDevice] = getFailedTestsForBuild(testResults["buildId"])[failedDevice]
```

Comment:
I think it's reasonably harder to follow what this code is doing now. Can you simplify this a bit? Ideas: 
- make the first `if exitStatus != 0 and numRetries > 0` a while loop, so you don't need to check these values multiple times
- extract some of this into a helper function

Reply:
I moved this chunk to a method, it's not 100% better, but I'm hoping it's clear enough as to what's going on now. But if not, let me know.